src/streaming/thread.pool.hh

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <condition_variable>
+#include <functional>
+#include <mutex>
+#include <optional>
+#include <queue>
+#include <string>
+#include <thread>
+
+namespace zarr {
+class ThreadPool
+{
+  public:
+    using JobT = std::function<bool(std::string&)>;
+
+    // The error handler `err` is called when a job returns false. This
+    // can happen when the job encounters an error, or otherwise fails. The
+    // std::string& argument to the error handler is a diagnostic message from
+    // the failing job and is logged to the error stream by the Zarr driver when
+    // the next call to `append()` is made.
+    ThreadPool(unsigned int n_threads,
+               std::function<void(const std::string&)> err);
+    ~ThreadPool() noexcept;
+
+    /**
+     * @brief Push a job onto the job queue.
+     *
+     * @param job The job to push onto the queue.
+     * @return true if the job was successfully pushed onto the queue, false
+     * otherwise.
+     */
+    [[nodiscard]] bool push_to_job_queue(JobT&& job);
```

Comment:
I suggest renaming it to `push_job` or `add_job`. There's no need to expose implementation details.